File: pastprograms/USACO/Silver/2018F-P2.py
import sys
sys.setrecursionlimit(100000)
def dfs(tile, boot):
    global boots
    global snow
    global possible
    global past
    global N
    global B
    if(tile >= N - 1): 
        possible = min(possible, boot)
        return
    if(boot >= B): return
    level, maxsnow, maxstep = snow[tile], boots[boot][0], boots[boot][1]
    if(maxsnow < level): return
    if((tile, boot) in past): return
    else: past.add((tile, boot))
    for i in range(tile + maxstep, tile, -1): 
        if(i >= N): 
            dfs(i, boot)
            return
        elif(snow[i] <= maxsnow):
            dfs(i, boot)
    for i in range(boot + 1, B): 
        if(snow[i] <= boots[i][0]): dfs(tile, i)

fin = open("snowboots.in", 'r')
fout = open("snowboots.out", 'w')
N, B = map(int, fin.readline().split(" "))
snow = fin.readline().split(" ")
boots = []
past = set()
for i in range(N): 
    snow[i] = int(snow[i])
for i in range(B): 
    depth, speed = map(int, fin.readline().split(" "))
    boots.append([depth, speed])
possible = B
dfs(0, 0)
fout.write(str(possible))
fin.close()
fout.close()

File: pastprograms/USACO/Silver/2018F-P2.cpp
#define pb push_back
#define mp make_pair
#define mt make_tuple
#define is insert
#define ll long long
#include <bits/stdc++.h>
using namespace std;

int N, M, a, b;
bool visited[250][250];
int snow[250];
pair<int, int> shoes[250];

void dfs (int pos, int boot) {
    if(pos >= (N - 1)) {
        visited[N - 1][boot] = true;
        return;
    }
    if(visited[pos][boot] || shoes[boot].first < snow[pos]) return;
    visited[pos][boot] = true;
    for (int i = boot + 1; i < M; i ++) {
        dfs(pos, i);
    }
    for (int i = pos + 1; i < (pos + shoes[boot].second + 1); i ++) {
        dfs(i, boot);
    }
}

int main() {
    ifstream fin; ofstream fout;
    fin.open("snowboots.in"); fout.open("snowboots.out");

    fin >> N >> M;
    for (int i = 0; i < N; i ++) fin >> snow[i];
    for (int i = 0; i < M; i ++) {
        fin >> a >> b;
        shoes[i] = mp(a, b);
    }
    dfs(0, 0);
    for (int i = 0; i <= M; i ++) {
        if(visited[N - 1][i]) {
            fout << i;
            return 0;
        }
    }
}